src/modules-load/modules-load.c

```diff
@@ -39,45 +88,249 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
+        _cleanup_fclose_ FILE *f = NULL;
+
+        f = fopen(FORMAT_PROC_FD_PATH(c->fd), "re");
+        if (!f) {
+                if (errno == ENOENT)
                         return 0;
 
-                return log_error_errno(r, "Failed to open %s: %m", path);
+                return log_error_errno(errno, "Failed to open %s: %m", c->original_path);
+        }
+
+        return apply_file(f, c->original_path, module_set);
+}
+
+static int do_direct_probe(OrderedSet *module_set) {
+        _cleanup_(sym_kmod_unrefp) struct kmod_ctx *ctx = NULL;
+        char *module;
+        int ret = 0, r;
+
+        r = module_setup_context(&ctx);
+        if (r < 0)
+                return log_error_errno(r, "Failed to initialize libkmod context: %m");
+
+        ORDERED_SET_FOREACH(module, module_set) {
+                r = module_load_and_warn(ctx, module, /* verbose = */ true);
+                if (r != -ENOENT)
+                        RET_GATHER(ret, r);
         }
 
-        log_debug("apply: %s", pp);
+        return ret;
+}
+
+static int enqueue_module_to_load(int sock, const char *module) {
+        ssize_t bytes;
+
+        assert(sock >= 0);
+        assert(module);
+
+        bytes = send(sock, module, strlen(module), 0);
+        if (bytes < 0)
+                return log_error_errno(errno, "Failed to send '%s' to thread pool: %m", module);
+
+        return 0;
+}
+
+static int dequeue_module_to_load(int sock, char *buffer, size_t buffer_len) {
+        ssize_t bytes;
+
+        assert(sock >= 0);
+        assert(buffer);
+
+        /* Dequeue one module to be loaded from the socket pair. In case no more
+         * modules are present, recv() will return 0. */
+        bytes = recv(sock, buffer, buffer_len, 0);
+        if (bytes == 0)
+                return 0;
+        if (bytes < 0)
+                return negative_errno();
+        if ((size_t)bytes == buffer_len)
+                return log_error_errno(SYNTHETIC_ERRNO(E2BIG), "Dequeued module name too long");
+
+        buffer[bytes] = '\0';
+
+        return 1;
+}
+
+static int run_prober(int sock) {
+        _cleanup_(sym_kmod_unrefp) struct kmod_ctx *ctx = NULL;
+        char buffer[MODULE_NAME_MAX_LEN + 1];
+        int ret = 0, r;
+
+        r = module_setup_context(&ctx);
+        if (r < 0)
+                return log_error_errno(r, "Failed to initialize libkmod context: %m");
+
         for (;;) {
-                _cleanup_free_ char *line = NULL;
-                int k;
+                r = dequeue_module_to_load(sock, buffer, sizeof(buffer));
+                if (ERRNO_IS_NEG_TRANSIENT(r))
+                        continue;
 
-                k = read_stripped_line(f, LONG_LINE_MAX, &line);
-                if (k < 0)
-                        return log_error_errno(k, "Failed to read file '%s': %m", pp);
-                if (k == 0)
+                if (r < 0)
+                        return log_error_errno(r, "Failed to receive from module queue: %m");
```

Comment:
because dequeue_module_to_load() logs on some errors but not on others, this might result in duplicate logging here.